transport/internet/ws/wsconn.go

```diff
@@ -0,0 +1,210 @@
+		This is very unlikely.
+	*/
+	readNext := func(b []byte) (n int, err error) { panic("Runtime unstable. Please report this bug to developer.") }
+
+	readNext = func(b []byte) (n int, err error) {
+		if ws.readBuffer == nil {
+			err = getNewBuffer()
+			if err != nil {
+				return 0, err
+			}
+		}
+
+		n, err = ws.readBuffer.Read(b)
+
+		if err == nil {
+			return n, err
+		}
+
+		if err == io.EOF {
+			ws.readBuffer = nil
+			if n == 0 {
+				return readNext(b)
+			}
+			return n, nil
+		}
+		return n, err
+
+	}
+	n, err = readNext(b)
+	ws.rlock.Unlock()
+	return n, err
+
+}
+
+func (ws *wsconn) Write(b []byte) (n int, err error) {
+	ws.wlock.Lock()
+	/*
+		process can crash as websocket report "concurrent write to websocket connection"
+		even if lock is persent.
+
+		This problem should have been resolved.
+	*/
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("WS workaround: recover", r)
+			ws.wlock.Unlock()
+		}
+	}()
+	if ws.connClosing {
+
+		return 0, io.EOF
+	}
+	writeWs := func(b []byte) (n int, err error) {
+		wr, err := ws.wsc.NextWriter(websocket.BinaryMessage)
+		if err != nil {
+			log.Warning("WS transport: ws connection NewFrameReader return " + err.Error())
+			ws.connClosing = true
+			ws.Close()
+			return 0, err
+		}
+		n, err = wr.Write(b)
+		if err != nil {
+			return 0, err
+		}
+		err = wr.Close()
+		if err != nil {
+			return 0, err
+		}
+		return n, err
+	}
+	n, err = writeWs(b)
+	ws.wlock.Unlock()
+	return n, err
+}
+func (ws *wsconn) Close() error {
+	ws.connClosing = true
+	ws.wlock.Lock()
+	ws.wsc.WriteControl(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""), time.Now().Add((time.Second * 5)))
+	ws.wlock.Unlock()
+	err := ws.wsc.Close()
+	return err
+}
+func (ws *wsconn) LocalAddr() net.Addr {
+	return ws.wsc.LocalAddr()
+}
+func (ws *wsconn) RemoteAddr() net.Addr {
+	return ws.wsc.RemoteAddr()
+}
+func (ws *wsconn) SetDeadline(t time.Time) error {
+	return func() error {
+		errr := ws.SetReadDeadline(t)
+		errw := ws.SetWriteDeadline(t)
+		if errr == nil || errw == nil {
+			return nil
+		}
+		if errr != nil {
+			return errr
+		}
+
+		return errw
+	}()
+}
+func (ws *wsconn) SetReadDeadline(t time.Time) error {
+	return ws.wsc.SetReadDeadline(t)
+}
+func (ws *wsconn) SetWriteDeadline(t time.Time) error {
+	return ws.wsc.SetWriteDeadline(t)
+}
+
+func (ws *wsconn) setup() {
+	ws.connClosing = false
+
+	ws.rlock = &sync.Mutex{}
+	ws.wlock = &sync.Mutex{}
+
+	ws.pingPong()
+}
+
+func (ws *wsconn) Reusable() bool {
+	return ws.reusable && !ws.connClosing
+}
+
+func (ws *wsconn) SetReusable(reusable bool) {
+	if !effectiveConfig.ConnectionReuse {
+		return
+	}
+	ws.reusable = reusable
+}
+
+func (ws *wsconn) pingPong() {
+	pongRcv := make(chan int, 0)
+	ws.wsc.SetPongHandler(func(data string) error {
+		pongRcv <- 0
+		return nil
+	})
+
+	go func() {
+		for !ws.connClosing {
+			ws.wlock.Lock()
+			ws.wsc.WriteMessage(websocket.PingMessage, nil)
```

Comment:
I think you are writing Ping messages too often. Try to sleep a while between each ping messages.